multiple-servers/README.md

```diff
@@ -0,0 +1,690 @@
+
+Timebox: 5 days
+
+Learning objectives:
+
+- Basic microservices ideas, separating concerns of services
+- Configure apache to talk to 2-3 copies of the API server
+- Some web security ideas (CORS)
+- Reverse proxy configuration, routing on path
+- Health checks
+
+In future:
+
+- Running applications in the cloud
+- Using a cloud-hosted databases
+- Multi-environment configuration
+
+## Project
+
+### Design
+
+In this project, we'll build something with the following architecture.
+
+![Architecture of this solution](./readme-assets/architecture.png)
+
+We'll make the file and API server, and use open source software called [nginx](https://nginx.org/) as the load balancer and router.
+
+You can follow the arrows to visualise the path the _request_ takes: an arrow from one box to another is getting data. The _response_ path is the arrow reversed.
+
+You will find the words "upstream" and "downstream" used too. Unfortunately this can be confusing because it depends if you are thinking about the request or response path. In general, upstream and downstream are thought of in terms of dependencies, or from the view of a response. So, the file and API servers are "upstream" of the load balancer: data _flows down the stream_ from the file server to the load balancer, and then the browser.
+
+Let's follow an example request, to `http://localhost:8080/index.html`:
+
+1. The browser requests `http://localhost:8080/index.html`
+1. The load balancer is listening on this port and receives the HTTP request
+1. It looks at the path (`/index.html`) of the request and tried to match it against its configuration
+1. The request path does not match `/api/*`, which you can read as "slash api slash anything". The `*` is often called a "wildcard".
+1. The request _does_ match `/*` — "slash anything" — so it **routes** the request to the file server
+
+This selection of where to send the request is called routing, and the order we perform the checks matters. `/*` matches everything, so if we checked that first, we'd never send any traffic to our API server. We check in order from the _most specific_ path pattern to the least specific path patterns.
+
+### Module & packages
+
+Our file layout for this project will look like this:
+
+```console
+api/
+    api.go
+assets/
+    ... website ...
+cmd/
+    static-server/
+        main.go
+    api-server/
+        main.go
+config/
+    nginx.conf
+static/
+    static.go
+go.mod
+```
+
+This is because we're building _two_ servers in the same module: `api` and `static`. Each has its own code and functionality.
+
+Each will be similar, but slightly different because one is connecting to a database and the other is serving files.
```

Comment:
Do we want to explain some ideas for _why_ you may split servers out? From this example, I don't think it's obvious why we have two services rather than one, and this seems like a good learning opportunity...

It feels like some of:
* Stateful vs stateless
* Contention for certain resources (e.g. the DB)
* Being able to out-source some (e.g. to a CDN)
* CPU vs disk vs network bound workloads (and specialised hardware)

may be relevant enough to explain now?

Reply:
Yes, agreed, thanks. I'll touch on a couple but not cover all.